### ISSUE TITLE
Block has wrong return value if it is reused by cascading math beforehand

### ISSUE THREAD

**Issue by VisualEhrmanntraut** [Vector35/binaryninja-api#6175]
**Version and Platform (required):**
- Binary Ninja Version: 4.3.6468-dev
- OS: macOS
- OS Version: 15.2 Beta (24C5089c)
- CPU Architecture: M3

**Bug Description:**
See below:
<img width="365" alt="Screenshot 2024-11-22 at 09 43 10" src="https://github.com/user-attachments/assets/688b4e9a-737f-4d4d-abb5-94adce9b62b6">
<img width="396" alt="Screenshot 2024-11-22 at 09 43 25" src="https://github.com/user-attachments/assets/dd7cab73-f288-491e-bd17-d742a6955df2">
Result:
<img width="708" alt="Screenshot 2024-11-22 at 09 43 31" src="https://github.com/user-attachments/assets/3c591085-8a7a-41bc-8dfe-042418489f1e">

**Expected Behavior:**
Should be `return 2`

**Comment by xusheng6** [comment]
@VisualEhrmanntraut Any chance you can share the binary? Sorry that I am not connecting the dots

**Comment by VisualEhrmanntraut** [comment]
@xusheng6 No, sorry, but I can share any further information needed. Just tell me what you need.

**Comment by xusheng6** [comment]
> @xusheng6 No, sorry, but I can share any further information needed. Just tell me what you need.

Sure, that is totally fine.

Can you show me the LLIL and MLIL as well? I am having some difficulty understanding what is going on

**Comment by VisualEhrmanntraut** [comment]
@xusheng6 Here,
LLIL
<img width="604" alt="Screenshot 2024-11-26 at 14 34 17" src="https://github.com/user-attachments/assets/a8c2b595-c859-44c6-a076-fb9722473213">
<img width="259" alt="Screenshot 2024-11-26 at 14 34 24" src="https://github.com/user-attachments/assets/39dbbfc9-8b3e-4349-afe4-59da05d240cc">

MLIL
<img width="577" alt="Screenshot 2024-11-26 at 14 34 53" src="https://github.com/user-attachments/assets/94f56148-931b-4976-8f6b-274dfaaa6094">
<img width="233" alt="Screenshot 2024-11-26 at 14 34 59" src="https://github.com/user-attachments/assets/83098807-9362-4c4a-9ea2-637964a00e12">

**Comment by xusheng6** [comment]
So it seems to me the switch-case is using `r8` as the control variable. However, I see it is compared against 0x2, which indicates some oddly looking code like this:

```
case 0x2:
    return 0x2;
```

where does the 0xb0007 come from?

**Comment by VisualEhrmanntraut** [comment]
It's a very weird function
<img width="629" alt="Screenshot 2024-11-26 at 14 39 26" src="https://github.com/user-attachments/assets/f52c5123-2df8-4f6b-94e0-ee5c5fb6ee3b">

**Comment by VisualEhrmanntraut** [comment]
essentially what this function does is convert a version number to an enumerator
i.e. 11.0.5 -> 1, 11.0.7 -> 2, etc

**Comment by xusheng6** [comment]
OK I now see how it works. Where is the `result` variable stored?

**Comment by xusheng6** [comment]
The code uses cascaded subtractions to take values off the incoming r8 register, and return rax values based on that. It seems we somehow screw up the return value tracking

**Comment by VisualEhrmanntraut** [comment]
@xusheng6 eax, afaict
<img width="721" alt="Screenshot 2024-11-26 at 14 57 25" src="https://github.com/user-attachments/assets/3c59125f-8f2b-4b71-b553-b0b4f39e3116">
<img width="568" alt="Screenshot 2024-11-26 at 14 57 37" src="https://github.com/user-attachments/assets/11b60bf6-f31c-45b5-a140-24b9fd600635">

**Comment by xusheng6** [comment]
Also @VisualEhrmanntraut , what about other cases after the one which should return 0x2? I see there should be quite a few more of them. Do we deal with it correctly?

**Comment by VisualEhrmanntraut** [comment]
@xusheng6 All of them seem correct except two, the "2" one and this
<img width="236" alt="Screenshot 2024-11-26 at 14 59 45" src="https://github.com/user-attachments/assets/3d3821dc-8dfd-4ac3-9db4-39ab54609f99">
that one should be `return 0x10`

Full function:
<img width="784" alt="Screenshot 2024-11-26 at 15 00 24" src="https://github.com/user-attachments/assets/505bfbe7-5fd7-4425-94b6-495117816fb5">

**Comment by xusheng6** [comment]
I understand what is happening, but do not know why it is happening. Will need someone else to look into it

**Comment by SlidyBat** [comment]
I think this may be a duplicate of this issue: https://github.com/Vector35/binaryninja-api/issues/6101

**Comment by VisualEhrmanntraut** [comment]
Hm, maybe. I didn't really know what to look for when creating the issue (or look through 1.5k issues).

**Comment by xusheng6** [comment]
v35 folks should search for `silent piano galaxy brick forest` to find the binary

**Comment by xusheng6** [comment]
With https://github.com/Vector35/binaryninja-api/issues/6101 fixed, this now looks like this:

![Image](https://github.com/user-attachments/assets/2ba6dc4c-44f4-4c66-9688-3c7d4ce3ccbc)

The only imperfection is `return zx.q(r8_4 - 0xcfffb)`, but the semantics is correct (r8_4 is known to be 0xd000b, and the expression value gives rise to 0x10). This is a separate issue (seems to be https://github.com/Vector35/binaryninja-api/issues/5638) and will be tracked otherwise
